Add ERC: PDAs (Programmatically Derived Addresses) (#890)
ethereum/ERCs: ERCS/erc-7879.md

---
EIP: 7879
Title: Program-Derived Addresses (PDAs) 

> Author: John Crunch, @JohnCrunch

Reviewer: ```suggestion
author: John Crunch (@JohnCrunch)
```

Discussions-to: TBD
Status: Draft
Type: Standards Track
Category: ERC
Created: 2025-2-08
---

## Abstract

> This Ethereum Improvement Proposal (EIP) introduces Program-Derived Addresses (PDAs) as a mechanism to generate deterministic, contract-derived addresses without private keys. Inspired by Solana’s PDA architecture, this standard enables smart contracts to create verifiable and immutable addresses using predefined seeds and cryptographic operations. PDAs provide stateless, permissionless interactions with smart contracts while ensuring security and uniqueness.

Reviewer: ```suggestion
Program-Derived Addresses (PDAs) are deterministic, contract-derived addresses without private keys. Inspired by Solana’s PDA architecture, this standard enables smart contracts to create verifiable and immutable addresses using predefined seeds and cryptographic operations. PDAs provide stateless, permissionless interactions with smart contracts while ensuring security and uniqueness.
```

> This Ethereum Improvement Proposal (EIP) introduces Program-Derived Addresses (PDAs) as a mechanism to generate deterministic, contract-derived addresses without private keys. Inspired by Solana’s PDA architecture, this standard enables smart contracts to create verifiable and immutable addresses using predefined seeds and cryptographic operations. PDAs provide stateless, permissionless interactions with smart contracts while ensuring security and uniqueness.

Reviewer: Your abstract is decent, but it's missing a high-level (but still technical) overview of _how_ the proposal accomplishes its goals. You should sketch out what the derivation mechanism is, at least briefly.

- **Deterministic Contract Addresses**: Smart contracts can generate specific, predictable addresses for interactions without requiring deployment.
- **Improved UX & Statelessness**: Eliminates the need for EOAs for specific interactions, reducing gas costs and simplifying account recovery.
- **Enhanced Security**: No private key exists for PDAs, mitigating risks of key exposure and hacks.
- **Trustless Multi-Contract State Management**: Contracts can interact with off-chain and on-chain data using derived addresses without requiring intermediary registries.
sdf
## Specification
A **Program-Derived Address (PDA)** is an Ethereum address deterministically generated by a contract using a combination of:

- A **base contract address**
- A **set of seed values**
- A **cryptographic derivation function** (Keccak256, SHA256, or other secure hashing algorithms)

The PDA is computed as follows:
```solidity
function derivePDA(address baseContract, bytes32[] memory seeds) public pure returns (address) {
    bytes32 hash = keccak256(abi.encodePacked(baseContract, seeds));
    return address(uint160(uint256(hash)));

> }

Reviewer 2: I have some concerns/questions about this since there is no Discussions link to EthMagicians.
This function derives an Ethereum address by truncating a 32-byte Keccak-256 hash to 20 bytes?. If so, this is not the standard method for deriving Ethereum addresses. Usually , an Ethereum address is obtained by taking the last 20 bytes of the Keccak-256 hash of the public key.
Solidity's abi.encodePacked performs tight packing, this makes variable-length arrays like seeds might not be properly hashed as intended.
No Error Handling or validation for invalid inputs.

#### 3. Trustless Escrow & Sub-Accounts
Protocols can create predictable, verifiable addresses for escrow, lending, and financial primitives without requiring independent contract deployments.

#### 4. Off-Chain State Commitments
PDAs provide a deterministic address for off-chain data commitments, ensuring verifiability while reducing on-chain storage costs.

#### 5. Efficient Multi-Party Control Without Gas Costs
Ethereum’s multisig wallets require explicit transactions to modify ownership structures, incurring gas fees. PDAs enable multi-party control without the need for continuous storage updates, as authority can be derived programmatically.

#### 6. On-Chain Order Books and Marketplaces
Ethereum’s on-chain order books require storage mappings, making them expensive in gas costs. PDAs allow orders to be stored in derived addresses without bloating contract storage, similar to Solana’s Serum order book model.

#### 7. Dynamic On-Chain Games
PDAs can facilitate player state storage in decentralized games, reducing reliance on smart contract storage while maintaining programmatic access to game states.

#### 8. Precomputed Addresses for DeFi and Oracles

> Lending protocols, AMMs, and on-chain oracles can generate deterministic PDAs for vaults, price feeds, and liquidity pools without requiring separate contract deployments, improving gas efficiency.

Reviewer: Use cases belong in Motivation. The Rationale section is for explaining choices made within the document itself.
